File: rust/core/src/timeline/placement/compatibility.rs
//! Element-to-track type compatibility mapping and validation.
//! Determines which track type a given element belongs on and
//! validates that placements are compatible.

/// Maps an element type to the track type it belongs on.
/// e.g. "audio" → "audio", "image" → "video", "sticker" → "graphic".
pub fn get_track_type_for_element_type(element_type: &str) -> &'static str {
    match element_type {
        "audio" => "audio",
        "text" => "text",
        "sticker" => "graphic",
        "graphic" => "graphic",
        "effect" => "effect",
        "video" => "video",
        "image" => "video",
        _ => "video", // Default fallback
    }
}

/// Returns true if an element of the given type can be placed on a track
/// of the given type.
pub fn can_element_go_on_track(element_type: &str, track_type: &str) -> bool {
    get_track_type_for_element_type(element_type) == track_type
}

/// The result of validating whether an element type is compatible with a
/// track type.
pub struct ValidationResult {
    /// Whether the element/track combination is valid.
    pub is_valid: bool,
    /// Description of the incompatibility, if any.
    pub error_message: Option<String>,
}

/// Validates that an element type is compatible with the given track type,
/// returning a `ValidationResult` with error details on mismatch.
pub fn validate_element_track_compatibility(
    element_type: &str,
    track_type: &str,
) -> ValidationResult {
    let is_valid = can_element_go_on_track(element_type, track_type);

    if !is_valid {
        return ValidationResult {
            is_valid: false,
            error_message: Some(format!(
                "{} elements cannot be placed on {} tracks",
                element_type, track_type
            )),
        };
    }

    ValidationResult {
        is_valid: true,
        error_message: None,
    }
}

File: apps/web/src/timeline/placement/compatibility.ts
/**
 * Element↔track compatibility checks and validation.
 *
 * Maps element types to track types and provides predicates for
 * determining whether an element can be placed on a given track.
 *
 * @module timeline/placement/compatibility
 */

import type { ElementType, TrackType } from "@/timeline";

const ELEMENT_TRACK_MAP: Record<ElementType, TrackType> = {
	audio: "audio",
	text: "text",
	sticker: "graphic",
	graphic: "graphic",
	effect: "effect",
	video: "video",
	image: "video",
};

/**
 * Returns the track type that accommodates the given element type.
 */
export function getTrackTypeForElementType({
	elementType,
}: {
	elementType: ElementType;
}): TrackType {
	return ELEMENT_TRACK_MAP[elementType];
}

/**
 * Returns `true` if the element type can be placed on the given track type.
 */
export function canElementGoOnTrack({
	elementType,
	trackType,
}: {
	elementType: ElementType;
	trackType: TrackType;
}): boolean {
	return getTrackTypeForElementType({ elementType }) === trackType;
}

/**
 * Validates element/track compatibility, returning an error message
 * when incompatible.
 */
export function validateElementTrackCompatibility({
	element,
	track,
}: {
	element: { type: ElementType };
	track: { type: TrackType };
}): { isValid: boolean; errorMessage?: string } {
	const isValid = canElementGoOnTrack({
		elementType: element.type,
		trackType: track.type,
	});

	if (!isValid) {
		return {
			isValid: false,
			errorMessage: `${element.type} elements cannot be placed on ${track.type} tracks`,
		};
	}

	return { isValid: true };
}
